File: 2019/day01/part2.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func getLines(s string) []int {
	stringLines := strings.Split(s, "\n")
	var numberLines []int

	for _, line := range stringLines {
		if len(line) == 0 {
			continue
		}
		n, err := strconv.Atoi(line)
		if err != nil {
			fmt.Printf("ERROR: could not convert string %+v to int", line)
			os.Exit(1)
		}

		numberLines = append(numberLines, n)
	}

	return numberLines
}

func openFile(filePath string) string {
	fileData, err := os.ReadFile(filePath)
	if err != nil {
		fmt.Printf("ERROR: could not read file %+v: %+v", filePath, err)
		os.Exit(1)
	}
	return string(fileData)
}

func getFuelRequiredForLine(startNumber int) int {
	if startNumber <= 0 {
		return 0
	}

	return startNumber + getFuelRequiredForLine((startNumber/3)-2)
}

func part2() int {
	fileData := openFile("./real-input.txt")
	lines := getLines(fileData)
	fuelRequired := 0

	for _, line := range lines {
		fuelRequired += getFuelRequiredForLine((line / 3) - 2)
	}
	return fuelRequired
}

File: 2024/day01/part2.js
const fs = require("fs");

fs.readFile("./real-input.txt", "utf8", (err, data) => {
  const lines = data.split("\n");
  const leftIds = [];
  const rightIds = [];
  lines
    .filter(Boolean)
    .map((line) => line.split("   ").map((n) => parseInt(n)))
    .forEach(([l, r]) => {
      leftIds.push(l);
      rightIds.push(r);
    });

  let res = 0;

  const cache = {};

  for (let l of leftIds) {
    if (cache.hasOwnProperty(l)) {
      res += cache[l];
      continue;
    }

    let c = 0;

    for (let r of rightIds) {
      if (l === r) c++;
    }

    const m = l * c;

    cache[l] = m;

    res += m;
  }

  console.log("result: ", res);
});
